autoflow/autoflow/date_stencil.py

```diff
@@ -0,0 +1,200 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+"""
+Defines the DateStencil class, to represent a pattern of dates relative to a reference date.
+"""
+
+import datetime
+from typing import Iterable, List, Optional, Sequence, Set, Tuple, Union
+
+import pendulum
+
+
+class InvalidDateIntervalError(ValueError):
+    """
+    Custom error to raise if a date stencil contains a date interval with start_date > end_date.
+    """
+
+    pass
+
+
+class DateStencil:
+    """
+    A class that represents a sequence of date intervals, which can be a
+    mixture of absolute dates and offsets relative to a reference date.
+
+    Parameters
+    ----------
+    raw_stencil : sequence of date, int and/or pairs of date/int
+        List of elements defining date intervals.
+        Each element can be:
+            - a date object corresponding to an absolute date,
+            - an int corresponding to an offset (in days) relative to a reference date,
+            - a length-2 sequence [start, end] of dates or offsets,
+              corresponding to a date interval (inclusive of both limits).
+    """
+
+    def __init__(
```

Comment:
I think a better way of passing these params and creating the object can be found. Not vital right now, but this is a little awkward.

Reply:
Agreed. Originally I was using the 'raw' format throughout, before I added the `DateStencil` class, and I think a little more refactoring is still needed here.